### PR TITLE
Add support for Softmax.

### PR DESCRIPTION
**Pull request by navahgar**
This PR adds support for Softmax in NNC.

### CODE REVIEW

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 1]
```diff
@@ -523,5 +523,231 @@ void testKernelSumMultipleAxes() {
...
+        %2 : int = prim::Constant[value=7]()
+        %3 : Tensor = aten::softmax(%0, %1, %2)
+        return (%3))IR";
+
+  auto a = iotaTensor({5, 3}, TensorOptions(kCPU).dtype(at::kFloat));
+
+  const std::string& verification_template =
+      R"IR(
+        # CHECK: for (int i0 = 0; i0 < 5
+        # CHECK: for (int i1 = 0; i1 < 3
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_kernel.cpp:542]: Nit: if lines are expected to be next to each other, it might be better to use `CHECK-NEXT`.

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me! Might be worth asking @nickgg to double check the reduction indices computation logic.

**Comment by nickgg** [comment]
Looks good to me, I think we should actually check the softmax math in the test though.

File: test/test_tensorexpr.py [hunk 2]
```diff
@@ -1065,6 +1066,41 @@ def easy(x, y):
...
+            return a + b + c + d
+
+        def test_ref(x, y):
+            a = special.softmax(x, axis=0)
+            b = special.softmax(y, axis=0)
+            c = special.softmax(x, axis=1)
+            d = special.softmax(y, axis=1)
+            return a + b + c + d
+
+        traced = torch.jit.trace(test, (torch.randn(2,3), torch.rand(2,3)))
```
Comment (ZolotukhinM) [inline test/test_tensorexpr.py:1085]: You probably need to specify `device=device` in inputs used for tracing too.

File: test/test_tensorexpr.py [hunk 3]
```diff
@@ -1065,6 +1066,41 @@ def easy(x, y):
...
+        inp = torch.randn(2, 3, device=device)
+        res = traced(inp, inp)
+        ref = traced(inp, inp)
```
Comment (ZolotukhinM) [inline test/test_tensorexpr.py:1088]: `res` and `ref` seem to be produced in the same way. Did you mean to call `test_ref`?

**Review by ZolotukhinM — COMMENTED** [review]
You probably also need to allow-list the softmax op in the fuser in order to get it to fusion groups:
https://github.com/pytorch/pytorch/blob/18876b572275fce59c284d41a9b486ed1cd02bc8/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L150-L153

File: test/test_tensorexpr.py (hunk 3 above)
Comment (navahgar) [reply test/test_tensorexpr.py]: Fixed.

File: test/test_tensorexpr.py (hunk 2 above)
Comment (navahgar) [reply test/test_tensorexpr.py]: Done.

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 1 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_kernel.cpp]: Done.

**Comment by navahgar** [comment]
Fixed the tests to compare and check with the actual softmax values. Also, enabled softmax in the fuser.

File: test/test_tensorexpr.py [hunk 4]
```diff
@@ -1065,6 +1066,43 @@ def easy(x, y):
...
+    def _test_softmax(self, device):
+        @torch.jit.script
+        def test(x, y):
+            a = F.softmax(x, dim=0)
+            b = F.softmax(y, dim=0)
+            c = F.softmax(x, dim=1)
+            d = F.softmax(y, dim=1)
+            return a + b + c + d
+
+        def test_ref(x, y):
```
Comment (bertmaher) [inline test/test_tensorexpr.py:1078]: This PR reminds me that we should stop using numpy/scipy as a reference and use torch eager mode instead...

Comment (navahgar) [reply test/test_tensorexpr.py:1078]: What is the recommendation / expectation of the community? Isn't it expected that our ops try to match the semantics of those in numpy / scipy? If thats the case, wouldn't it better to stick to them as the reference points?

**Review by nickgg — APPROVED** [review]
Let's do it! Behind the reduction support guard ofc.

File: test/test_tensorexpr.py (hunk 4 above)
Comment (ZolotukhinM) [reply test/test_tensorexpr.py:1078]: If we diverge from eager mode, then it's a bug in our code and we should fix it. If the eager mode diverges from numpy/scipy, we should raise this issue, but it's not always a must-fix bug, and it is not blocking *TE* changes.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 5]
```diff
@@ -1554,6 +1563,64 @@ Tensor* TensorExprKernel::computeSum(const torch::jit::Value* v) {
...
+  return Compute("aten_softmax", output_dims, [&](ParameterList& indices) {
+    // The sum tensor uses all indices other than the softmax dimension
+    // index.
+    std::vector<ExprHandle> new_indices;
+    for (size_t i = 0; i < indices.size(); ++i) {
+      if (i != softmax_dim) {
+        new_indices.push_back(indices[i]);
+      }
+    }
+    return e->call(indices) / sum->call(new_indices);
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1620]: In softmax, both the sum and max need to be computed. All elements should be shifted with the max first. Otherwise, if all your elements are quite large, it quickly leads to overflow. 

The tricky part is that for performant execution, the max and sum need to computed together.

Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1620]: The default lowering should inline e->call(...). Either call an inline directly, or just manually. Either way is fine with me.

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1620]: Why do we need to compute max here? IIUC, softmax only requires the sum over exp of all the elements along the given dimension. Am I missing something?

I understand that the sum could lead to overflow quickly. But that is a known issue with softmax, right? Isn't that one of the reasons we have `log_softmax`?

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1620]: Are you suggesting I inline `e->call(...)` here? Why is that necessary?

Comment (pearu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1621]: @navahgar 
> Why do we need to compute max here? IIUC, softmax only requires the sum over exp of all the elements along the given dimension. Am I missing something?

Yes, the definition of softmax contains only the sum of exp of all the elements along the given dimensions.
However, subtracting the max of the elements before applying exponent gives a numerically stable algorithm while the output will match with the softmax definition.

> I understand that the sum could lead to overflow quickly. But that is a known issue with softmax, right? Isn't that one of the reasons we have `log_softmax`?

Right, it is a known issue and it has a know solution as described above. Also the computation of log_softmax uses the max-substraction-trick.

@zheng-xq 
When implementing stable and performant softmax, you should follow the torch softmax implementation:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L28-L62

Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1621]: We can't call inline without constructing a Loopnest (which we don't want to do here) - the only choice we have is to inline manually.

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1621]: Thanks for the explanation @pearu.

**Comment by nickgg** [comment]
Correctness is important to get right up front, but performance we can improve iteratively. Once we fix the heuristic to inline into reductions that call should get inlined automatically, I think.

I think we should land and keep working on reduction perf in general.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 5 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: As discussed offline, this is currently not possible with the infrastructure we have. But I am planning to fix inlining to handle these cases, in a separate PR.

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Added the max-subtraction-trick to the implementation.

File: test/test_tensorexpr.py (hunk 4 above)
Comment (navahgar) [reply test/test_tensorexpr.py]: Changed it to use eager mode as ref.

**Comment by nickgg** [comment]
@zheng-xq @pearu - addressed your concerns?

**Review by zheng-xq — APPROVED** [review]
